text/0000-maybe-dangling.md

```diff
@@ -0,0 +1,268 @@
+
+However, note that `thread` continues to run even after `signal_done`! Now consider what happens if the closure captures a reference of lifetime `'lifetime`:
+- The type of `closure` is a struct (the implicit unnameable closure type) with a `&'lifetime mut T` field.
+  References passed to a function must be live for the entire duration of the call.
+- The closure runs, `signal_done` runs.
+  Then -- potentially -- this thread gets scheduled away and the main thread runs, seeing the signal and returning to the user.
+    Now `'lifetime` ends and the memory the reference points to might be deallocated.
+- Now we have UB! The reference that as passed to `thread` with the promise of remaining live for the entire duration of the function, actually got deallocated while the function still runs. Oops.
+
+### Example 3
+
+As a third example, consider a type that wants to store a "pointer together with some data borrowed from that pointer", like the `owning_ref` crate. This will usually boil down to something like this:
+
+```rust
+unsafe trait StableDeref: Deref {}
+
+struct OwningRef<U, T: StableDeref<Target=U>> {
+    buffer: T,
+    ref_: NonNull<U>, // conceptually borrows from `buffer`.
+}
+```
+
+Such a type is unsound when `T` is `&mut U` or `Box<U>` because those types are assumed by the compiler to be unique, so any time `OwningRef` is passed around, the compiler can assume that `buffer` is a unique pointer -- an assumption that this code breaks because `ref_` points to the same memory!
+
+### Goal of this RFC
+
+The goal of this RFC is to
+- make the first example UB-free without code changes
+- make the second example UB-free without needing to add `unsafe` code
+- make it possible to define a type like the third example
+
+(Making the 2nd example UB-free without code changes would incur cost across the ecosystem, see the alternatives discussed below.)
+
+The examples described above are far from artificial, here are some real-world crates that need `MaybeDangling` to ensure their soundness (some currently crudely work-around that problem with `MaybeUninit` but that is really not satisfying):
+- [Yoke](https://github.com/unicode-org/icu4x/issues/3696) and [Yoke again](https://github.com/unicode-org/icu4x/issues/2095) (the first needs opting-out of `dereferenceable` for the yoke, the latter needs opting-out of `noalias` for both yoke and cart)
+- [ouroboros](https://github.com/joshua-maros/ouroboros/issues/88)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+To handle situations like this, Rust has a special type called `MaybeDangling<P>`:
+references and boxes in `P` do *not* have to be dereferenceable or follow aliasing guarantees.
+This applies inside nested references/boxes inside `P` as well.
+They still have to be non-null and aligned, and it has to at least be *possible* that there exists valid data behind that reference (i.e., `MaybeDangling<&!>` is still invalid).
```

Comment:
> `MaybeDangling<&!>` is still invalid

Based on my current understanding of ref-to-never, this can change to "`MaybeDangling<&!>` has an always-violated safety invariant"

Reply:
`&!` has an impossible *validity* invariant according to the Reference, Miri, and MiniRust. So I don't think the text should change.